netbox/utilities/jinja2.py

```diff
@@ -29,10 +29,16 @@ def get_source(self, environment, template):
 
         # Find and pre-fetch referenced templates
         if referenced_templates := find_referenced_templates(environment.parse(template_source)):
-            self.cache_templates({
-                df.path: df.data_as_string for df in
-                DataFile.objects.filter(source=self.data_source, path__in=referenced_templates)
-            })
+            if None in referenced_templates:
+                self.cache_templates({
+                    df.path: df.data_as_string for df in
+                    DataFile.objects.filter(source=self.data_source)
+                })
+            else:
+                self.cache_templates({
+                    df.path: df.data_as_string for df in
+                    DataFile.objects.filter(source=self.data_source, path__in=referenced_templates)
+                })
```

Comment:
We can conditionally apply the `path__in` filter to avoid duplicating code. We should also explain the significance of None in the paths list as its meaning is not obvious IMO.
```suggestion
            related_files = DataFile.objects.filter(source=self.data_source)
            # None indicates the use of dynamic resolution. If dependent files are statically
            # defined, we can filter by path for optimization.
            if None not in referenced_templates:
                related_files = related_files.filter(path__in=referenced_templates)
            self.cache_templates({
                df.path: df.data_as_string for df in related_files
            })
```

Reply:
Django querysets are lazy; they won't actually trigger a database query until evaluation has been forced (when the dict is generated).